# Data Exploration

## Data Import

In [22]:
import numpy as np
import pandas as pd
import sys

In [23]:
data_folder = "data/"

    SerialID  building_id  zone_id     SEEB  \
0  311487854            0        0  2-1-3-0   
1  311235119            0        0  2-1-3-0   
2  311235135            0        0  2-1-3-0   
3  311235144            0        0  2-1-3-0   
4  311488415            0        0  2-1-3-0   

                                    S_text  \
0        Fjernvarme uden veksler - direkte   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4        Fjernvarme uden veksler - direkte   

                                              L_text                Type  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
1  Ejendommen opvarmes med fjernvarme, og anlægge...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
4  Bygningen opvarmes med fjernvarme. Anlægget er...  Direct

In [24]:
d_hjorring = pd.read_excel(
    data_folder + "data_district_heating.xlsx", sheet_name="Hjorring"
)
d_favrskov = pd.read_excel(
    data_folder + "data_district_heating.xlsx", sheet_name="Favrskov"
)
d_aarhus = pd.read_excel(
    data_folder + "data_district_heating.xlsx", sheet_name="Aarhus"
)
d_gt = pd.read_excel(
    data_folder + "data_district_heating.xlsx", sheet_name="Ground Truth"
)
print(d_gt.head())

    SerialID  building_id  zone_id     SEEB  \
0  311239024            0        0  2-1-3-0   
1  311239047            0        0  2-1-3-0   
2  311239049            0        0  2-1-3-0   
3  311244995            0        0  2-1-3-0   
4  311251256            0        0  2-1-3-0   

                                    S_text  \
0  Fjernvarme uden veksler (direkte anlæg)   
1  Fjernvarme uden veksler (direkte anlæg)   
2  Fjernvarme uden veksler (direkte anlæg)   
3  Fjernvarme uden veksler (direkte anlæg)   
4  Fjernvarme uden veksler (direkte anlæg)   

                                              L_text                Type  \
0  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
1  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
2  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
3  Bygningen opvarmes med fjernvarme. Anlægget er...  DirectDistrictHeat   
4  :Klynge A, B, C og D:\nBygningerne opvarmes me...  Direct

## Data Size

In [27]:
print("Data Favrskov shape: ", d_favrskov.shape)
print("Data Hjorring shape: ", d_hjorring.shape)
print("Data Aarhus shape: ", d_aarhus.shape)
print("Data Ground Truth shape: ", d_gt.shape)

Data Favrskov shape:  (111, 15)
Data Hjorring shape:  (6138, 15)
Data Aarhus shape:  (29542, 15)
Data Ground Truth shape:  (2272, 42)


Let's merge the dataset on the SerialID to check whether the ground truth data are present in the rest of the samples.

In [28]:
d_f_gt = pd.merge(left=d_favrskov, right=d_gt, on="SerialID")
d_h_gt = pd.merge(left=d_hjorring, right=d_gt, on="SerialID")
d_a_gt = pd.merge(left=d_aarhus, right=d_gt, on="SerialID")
print(d_f_gt.shape)
print(d_h_gt.shape)
print(d_a_gt.shape)

(0, 56)
(0, 56)
(0, 56)


From the previous result, we can say that the ground truth is not present in data.
Thus we have 2272 labeled samples.

In [29]:
print(d_favrskov.shape[0] + d_hjorring.shape[0] + d_aarhus.shape[0])

35791


On the other hand, we have 35791 entries we want to label.

## Features of the Ground Truths

In [30]:
print(d_gt[d_gt["SerialID"] == 311795931].iloc[0, 0:31])

SerialID                                                      311795931
building_id                                                           0
zone_id                                                               0
SEEB                                                            2-1-3-0
S_text                                              Fjernvarmeforsyning
L_text                Forsyningsformen er fjernvarme som leveres af ...
Type                                          DistrictHeatWithExchanger
Pieces1                                                             2.0
Manufacturer1                                                      Reci
SubType1                                                          VT 90
HxType1                                                      Rørveksler
NominelEffectEach1                                300 kW (total 600 kW)
Year1                                                              1984
Pieces2                                                         

In [31]:
d_gt_labels = d_gt.iloc[:, 7:31]
print(d_gt_labels.columns)
print("Number of lables: ", d_gt_labels.shape[1])

Index(['Pieces1', 'Manufacturer1', 'SubType1', 'HxType1', 'NominelEffectEach1',
       'Year1', 'Pieces2', 'Manufacturer2', 'SubType2', 'HxType2',
       'NominelEffectEach2', 'Year2', 'Pieces3', 'Manufacturer3', 'SubType3',
       'HxType3', 'NominelEffectEach3', 'Year3', 'Pieces4', 'Manufacturer4',
       'SubType4', 'HxType4', 'NominelEffectEach4', 'Year4'],
      dtype='object')
Number of lables:  24


In [32]:
print(100 - d_gt_labels.isnull().mean() * 100)

Pieces1               100.000000
Manufacturer1         100.000000
SubType1               17.605634
HxType1                92.737676
NominelEffectEach1      7.438380
Year1                  41.153169
Pieces2                 4.225352
Manufacturer2           4.269366
SubType2                3.169014
HxType2                 3.917254
NominelEffectEach2      0.308099
Year2                   3.609155
Pieces3                 0.352113
Manufacturer3           0.352113
SubType3                0.220070
HxType3                 0.352113
NominelEffectEach3      0.088028
Year3                   0.308099
Pieces4                 0.264085
Manufacturer4           0.264085
SubType4                0.176056
HxType4                 0.264085
NominelEffectEach4      0.000000
Year4                   0.176056
dtype: float64


From the previous result, we know that every entry has informations about `Pieces1` and `Manufacturer1`.
We can see that `NominalEffectEach4` is always empty, and thus we do not have information about this column.
Finally, we can see that data are very sparse, indeed very few columns are filled.

In [33]:
null_counts = d_gt_labels.isnull().sum(axis=1)
missing_values = null_counts.value_counts()

missing_index = np.array(missing_values.index)
missing_index.sort()
missing_values = np.array(missing_values[missing_index])
missing_values_cum = np.cumsum(missing_values)
percentages = missing_values_cum / missing_values_cum[-1] * 100

for i in range(missing_index.shape[0]):
    print(
        "Number of samples: ",
        missing_values_cum[i],
        f"\tCorresponding percentage: {percentages[i]:.3f} No more than {missing_index[i]} entries missing",
    )

Number of samples:  1 	Corresponding percentage: 0.044 No more than 3 entries missing
Number of samples:  5 	Corresponding percentage: 0.220 No more than 6 entries missing
Number of samples:  6 	Corresponding percentage: 0.264 No more than 7 entries missing
Number of samples:  7 	Corresponding percentage: 0.308 No more than 8 entries missing
Number of samples:  8 	Corresponding percentage: 0.352 No more than 10 entries missing
Number of samples:  10 	Corresponding percentage: 0.440 No more than 12 entries missing
Number of samples:  61 	Corresponding percentage: 2.685 No more than 13 entries missing
Number of samples:  65 	Corresponding percentage: 2.861 No more than 14 entries missing
Number of samples:  85 	Corresponding percentage: 3.741 No more than 16 entries missing
Number of samples:  90 	Corresponding percentage: 3.961 No more than 17 entries missing
Number of samples:  164 	Corresponding percentage: 7.218 No more than 18 entries missing
Number of samples:  396 	Corresponding p

In [34]:
print(
    f"Average length of S_text: {d_gt['S_text'].str.len().mean():.2f}\t standard deviation: {d_gt['S_text'].str.len().std():.2f}"
)
print(
    f"Average length of L_text: {d_gt['L_text'].str.len().mean():.2f} standard deviation: {d_gt['L_text'].str.len().std():.2f}"
)

Average length of S_text: 48.65	 standard deviation: 14.20
Average length of L_text: 195.77 standard deviation: 175.14


# Export Gold Label in csv

In [36]:
d_gt["text"] = (
    d_gt["S_text"].fillna("").astype(str) + ". " + d_gt["L_text"].fillna("").astype(str)
)
print(
    "Total number of sentences: ", (d_gt["text"].str.count(r"[.!?](?:\s|$)") + 1).sum()
)

Total number of sentences:  11017


In [37]:
print("Total number of lables: ", d_gt_labels.count().sum())

Total number of lables:  8662


In [38]:
n_sentences = (d_gt["text"].str.count(r"[.!?](?:\s|$)") + 1).sum()
n_labels = d_gt_labels.count().sum()
print(
    "Labels on sentences ratio: ",
    n_labels / n_sentences,
    " Labels on sentences percentage: ",
    n_labels / n_sentences * 100,
)

Labels on sentences ratio:  0.7862394481256241  Labels on sentences percentage:  78.62394481256241


In [39]:
for col in d_gt_labels.columns:
    print(f"{col} has the following values: {d_gt_labels[col].value_counts()}")

Pieces1 has the following values: Pieces1
1.00    2022
2.00     184
3.00      39
4.00      22
5.00       2
0.25       1
0.53       1
0.94       1
Name: count, dtype: int64
Manufacturer1 has the following values: Manufacturer1
Unknown                           1452
Reci                               212
Gemina Termix                      106
None - direct district heating      80
CTC                                 67
Sondex Teknik                       66
Reflex                              44
Danfoss Redan                       28
Danfoss                             24
Elge                                22
Kähler & Breum                      21
WPH Teknik                          21
Metro Therm                         20
Cetetherm                           20
Alfa Laval                          19
Ajva                                14
Armatec                             13
Ajva                                 6
APV                                  5
Ducon                            

In [44]:
col = "HxType1"
print(f"{col} has the following values: {d_gt_labels[col].value_counts()}")

HxType1 has the following values: HxType1
Isoleret varmeveksler          1372
Fjernvarmeveksler               238
Varmeveksler                    148
Uisoleret varmeveksler           93
Isoleret pladevarmeveksler       87
Isoleret veksler                 80
Isoleret rørvarmeveksler         48
Isoleret rørveksler              12
Isoleret fjernvarmeveksler        8
Veksler                           5
Isoleret pladeveksler             5
Isoleret pladerørveksler          3
Pladeveksler                      2
Rørveksler                        2
Uisoleret pladevarmeveksler       2
Isoleret dampvarmeveksler         1
Pladevarmeveksler                 1
Name: count, dtype: int64


In [42]:
print(d_gt.loc[195, "text"])

Fjernvarme med ISOLERET veksler (indirekte anlæg). Bygningen opvarmes med fjernvarme via varmecentral placeret i kælderen Njalsgade 12.

Varmecentral - Njalsgade 12. 
Forsyner AB Egilshus - 4 bygninger.      
Anlægget er udført med 4 stk. isoleret varmevekslere og indirekte centralvarmevand i fordelingsnettet.
Fabrikat reflex, år 2007 (1 stk.) og Fabrikat Reci, år 1989 (3 stk.).




In [ ]:
for col in d_gt_labels.columns[:6]:
    print(col, "has", len(d_gt_labels[col].unique()), "unique fields")

In [ ]:
d_gt_labels["Manufacturer1"].unique()